File: SurvCop/R/survivalconditionalcdf.r
#' conditional CDF of the survival response given mixed discrete/continuous explanatory variables
#' 
#' @description
#' conditional CDF of the survival response given mixed discrete/continuous explanatory variables
#' 
#' @param u-score scalar or vector of the response variable of length n
#' @param paramvec vector of length (d-1) if all the copula families associated with the survival response variable are one-parameter families; vector of length (d-1)+ otherwise
#' @param param1_complete d-1 by d-1 matrix for the first parameter of the copula models for the predictors
#' @param param2_complete d-1 by d-1 matrix for the second parameter of the copula models for the predictors
#' @param fam d by d matrix for the copula families coded by the VineCopula package
#' @param A d by d matrix for the copula structure
#' @param data_cdf_plus n by (d-1) matrix for CDF from the right
#' @param data_cdf_minus n by (d-1) matrix for CDF from the left
#' @param is_disc vector of length d representing if the variable is discrete or not
#
#' @return the conditional CDF of the response variable given the explanatory variables of length n
#'
#' @import VineCopula
#' @export

survivalconditionalcdf = function(u, paramvec, param1_complete, param2_complete, fam, A, data_cdf_plus, data_cdf_minus, is_disc) {
	d = ncol(A)
	n = nrow(data_cdf_plus)
	
	# permute all arguments based on the order of A
	A_order = diag(A)
	A = varrayperm(A, order(A_order))
	
	data_cdf_plus = data_cdf_plus[, A_order[-d]]
	data_cdf_minus = data_cdf_minus[, A_order[-d]]
	
	if (n == 1) {
		data_cdf_plus = matrix(data_cdf_plus, nrow = 1)
		data_cdf_minus = matrix(data_cdf_minus, nrow = 1)
	}
	
	is_disc = is_disc[A_order]
	
	paramvec = parse_parameter(paramvec, fam[1:(d-1), d], d)
	param1_vec = paramvec[1:(d-1)]
	param2_vec = paramvec[d:length(paramvec)]
	
	parammat1 = matrix(0, nrow = d, ncol = d)
	parammat1[1:(d-1), 1:(d-1)] = param1_complete
	parammat1[1:(d-1), d] = param1_vec
	
	parammat2 = matrix(0, nrow = d, ncol = d)
	parammat2[1:(d-1), 1:(d-1)] = param2_complete
	parammat2[1:(d-1), d] = param2_vec
	
	M = matrix(0, nrow = d, ncol = d)
	for (i in 1:d) {
		for (j in i:d) {
			M[i, j] = max(A[1:i, j])
		}
	}
	
	I = matrix(0, nrow = d, ncol = d)
	for (i in 2:d) {
		for (j in i:d) {
			if (A[i, j] < M[i, j]) {
				I[i-1, M[i, j]] = 1
			}
		}
	}
	
	v_plus = rep(0, d)
	v_minus = rep(0, d)
	vp_plus = rep(0, d)
	vp_minus = rep(0, d)
	s_plus = rep(0, d)
	s_minus = rep(0, d)
	w_plus = rep(0, d)
	w_minus = rep(0, d)
	
	result = numeric(length(u))
	
	for (i in 1:length(u)) {
		data_cdf_plus_i = c(data_cdf_plus[i, ], u[i])
		data_cdf_minus_i = c(data_cdf_minus[i, ], u[i])
		
		# initialization steps
		for (j in 1:d) {
			s_plus[j] = data_cdf_plus_i[A[1, j]]
			s_minus[j] = data_cdf_minus_i[A[1, j]]
			w_plus[j] = data_cdf_plus_i[j]
			w_minus[j] = data_cdf_minus_i[j]
		}
		
		if (d == 2) {
			if (is_disc[A[d-1, d]]) {
				ucond = (BiCopCDF(s_plus[d], w_plus[d], fam[d-1, d], parammat1[d-1, d], parammat2[d-1, d]) - BiCopCDF(s_minus[d], w_plus[d], fam[d-1, d], parammat1[d-1, d], parammat2[d-1, d])) / (s_plus[d] - s_minus[d])
			} else {
				# condition on first variable (not discrete)
				ucond = BiCopHfunc2(w_plus[d], s_plus[d], fam[d-1, d], parammat1[d-1, d], parammat2[d-1, d])
			}
		} else {
			# levels 2:(d-1)
			for (l in 2:(d-1)) {
				s_plus[s_plus >= 1] = .999
				s_minus[s_minus >= .999] = .998
				w_plus[w_plus >= 1] = .999
				w_minus[w_minus >= .999] = .998
				s_plus[s_plus <= .001] = .002
				s_minus[s_minus <= 0] = .001
				w_plus[w_plus <= .001] = .002
				w_minus[w_minus <= 0] = .001
				for (j in l:d) {
					if (I[l-1, j] == 1) {
						if (is_disc[j]) {
							vp_plus[j] = (BiCopCDF(s_plus[j], w_plus[j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j]) - BiCopCDF(s_plus[j], w_minus[j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j])) / (w_plus[j] - w_minus[j])
							vp_minus[j] = (BiCopCDF(s_minus[j], w_plus[j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j]) - BiCopCDF(s_minus[j], w_minus[j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j])) / (w_plus[j] - w_minus[j])
						} else {
							# condition on second variable (not discrete) for vprime 
							vp_plus[j] = BiCopHfunc2(s_plus[j], w_plus[j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j])
							vp_minus[j] = BiCopHfunc2(s_minus[j], w_plus[j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j])
						}
					}
					if (is_disc[A[l-1, j]]) {
						v_plus[j] = (BiCopCDF(s_plus[j], w_plus[j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j]) - BiCopCDF(s_minus[j], w_plus[j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j])) / (s_plus[j] - s_minus[j])
						v_minus[j] = (BiCopCDF(s_plus[j], w_minus[j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j]) - BiCopCDF(s_minus[j], w_minus[j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j])) / (s_plus[j] - s_minus[j])
					} else {
						# condition on first variable (not discrete)
						v_plus[j] = BiCopHfunc2(w_plus[j], s_plus[j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j])
						v_minus[j] = BiCopHfunc2(w_minus[j], s_plus[j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j])
					}
				}
				
				for (j in (l+1):d) {
					if (A[l, j] == M[l, j]) {
						s_plus[j] = v_plus[M[l, j]]
						s_minus[j] = v_minus[M[l, j]]
					} else {
						s_plus[j] = vp_plus[M[l, j]]
						s_minus[j] = vp_minus[M[l, j]]
					}
					w_plus[j] = v_plus[j]
					w_minus[j] = v_minus[j]
				}
				
				if (l == (d - 1)) {
					if (is_disc[A[d-1, d]]) {
						ucond = (BiCopCDF(s_plus[d], w_plus[d], fam[d-1, d], parammat1[d-1, d], parammat2[d-1, d]) - BiCopCDF(s_minus[d], w_plus[d], fam[d-1, d], parammat1[d-1, d], parammat2[d-1, d])) / (s_plus[d] - s_minus[d])
					} else {
						# condition on first variable (not discrete)
						ucond = BiCopHfunc2(w_plus[d], s_plus[d], fam[d-1, d], parammat1[d-1, d], parammat2[d-1, d])
					}
				}
			}
		}
		result[i] = ucond
	}
	return(result)
}

File: SurvCop/src/survival_conditional_cdf.f90
subroutine survivalconditionalcdf(n, d, u, parammat1, parammat2, fam, A, datacdfplus, datacdfminus, isdisc, output)
  implicit none
  integer n, d, i, j, l
  integer fam(d,d), A(d,d), M(d,d), Imat(d,d)
  double precision parammat1(d,d), parammat2(d,d)
  double precision u(n), datacdfplus(n,d-1), datacdfminus(n,d-1), datacdfplusi(d), datacdfminusi(d), output(n)
  double precision vplus(d), vminus(d), vpplus(d), vpminus(d), splus(d), sminus(d), wplus(d), wminus(d)
  logical isdisc(d)
  double precision ucond, tmp, tmp1, tmp2, tmp3, tmp4
  !print*, n, d
  !print*, parammat1, parammat2
  !print*, fam
  !print*, (A(i,:), i=1,5)
  !print*, datacdfplus
  !print*, datacdfminus
  !print*, isdisc
  
  vplus=0.d0; vminus=0.d0; vpplus=0.d0; vpminus=0.d0; splus=0.d0; sminus=0.d0; wplus=0.d0; wminus=0.d0
  M=0; Imat = 0
  do i = 1,d
    do j = i,d
      M(i, j) = maxval(A(1:i, j))
    end do
  end do
  !print*, M
  
  do i = 2,d
    do j = i,d
      if (A(i, j) < M(i, j)) then
        Imat(i-1, M(i, j)) = 1
      end if
    end do
  end do
  !print*, Imat
  
  do i = 1,n
    datacdfplusi(1:(d-1)) = datacdfplus(i, :)
    datacdfplusi(d) = u(i)
    datacdfminusi(1:(d-1)) = datacdfminus(i, :)
    datacdfminusi(d) = u(i)
    do j = 1,d
      splus(j) = datacdfplusi(A(1, j))
      sminus(j) = datacdfminusi(A(1, j))
      wplus(j)= datacdfplusi(j)
      wminus(j) = datacdfminusi(j)
    end do
    
    if (d == 2) then
      if (isdisc(A(d-1, d))) then
        call bivcopcdf(splus(d), wplus(d), fam(d-1, d), parammat1(d-1, d), parammat2(d-1, d), tmp1)
        call bivcopcdf(sminus(d), wplus(d), fam(d-1, d), parammat1(d-1, d), parammat2(d-1, d), tmp2)
        ucond = (tmp1 - tmp2) / (splus(d) - sminus(d))
      else
        call bivcopccdf(wplus(d), splus(d), fam(d-1, d), parammat1(d-1, d), parammat2(d-1, d), tmp)
        ucond = tmp
      end if
    else
      do l = 2,d-1
        where (splus >= 1.d0)
          splus = .999d0
        end where
        where (sminus >= .999d0)
          sminus = .998d0
        end where
        where (wplus >= 1.d0)
          wplus = .999d0
        end where
        where (wminus >= .999d0)
          wminus = .998d0
        end where
        where (splus <= .001d0)
          splus = .002d0
        end where
        where (sminus <= 0.d0)
          sminus = .001d0
        end where
        where (wplus <= .001d0)
          wplus = .002d0
        end where
        where (wminus <= 0.d0)
          wminus = .001d0
        end where
        
        do j = 1,d
          if (Imat(l-1, j) == 1) then
            if (isdisc(j)) then
              call bivcopcdf(splus(j), wplus(j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp1)
              call bivcopcdf(splus(j), wminus(j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp2)
              vpplus(j) = (tmp1 - tmp2) / (wplus(j) - wminus(j))
              call bivcopcdf(sminus(j), wplus(j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp3)
              call bivcopcdf(sminus(j), wminus(j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp4)
              vpminus(j) = (tmp3 - tmp4) / (wplus(j) - wminus(j))
            else
              call bivcopccdf(splus(j), wplus(j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp1)
              vpplus(j) = tmp1
              call bivcopccdf(sminus(j), wplus(j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp2)
              vpminus(j) = tmp2
            end if
          end if
          if (isdisc(A(l-1, j))) then
            call bivcopcdf(splus(j), wplus(j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp1)
            call bivcopcdf(sminus(j), wplus(j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp2)
            vplus(j) = (tmp1 - tmp2) / (splus(j) - sminus(j))
            call bivcopcdf(splus(j), wminus(j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp3)
            call bivcopcdf(sminus(j), wminus(j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp4)
            vminus(j) = (tmp3 - tmp4) / (splus(j) - sminus(j))
          else
            call bivcopccdf(wplus(j), splus(j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp1)
            vplus(j) = tmp1
            call bivcopccdf(wminus(j), splus(j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp2)
            vminus(j) = tmp2
          end if
        end do
        
        do j = l+1,d
          if (A(l,j) == M(l,j)) then
            splus(j) = vplus(M(l,j))
            sminus(j) = vminus(M(l,j))
          else
            splus(j) = vpplus(M(l,j))
            sminus(j) = vpminus(M(l,j))
          end if
          wplus(j) = vplus(j)
          wminus(j) = vminus(j)
        end do
        
        if (l == (d - 1)) then
          if (isdisc(A(d-1, d))) then
            call bivcopcdf(splus(d), wplus(d), fam(d-1,d), parammat1(d-1,d), parammat2(d-1,d), tmp1)
            call bivcopcdf(sminus(d), wplus(d), fam(d-1,d), parammat1(d-1,d), parammat2(d-1,d), tmp2)
            ucond = (tmp1 - tmp2) / (splus(d) - sminus(d))
          else
            call bivcopccdf(wplus(d), splus(d), fam(d-1, d), parammat1(d-1, d), parammat2(d-1, d), tmp)
            ucond = tmp
          end if
        end if
      end do
    end if
    output(i) = ucond
  end do
  return
end